# Chapter 1: Variables, Types, and Expressions

---

## The CRAWL → WALK → RUN Framework

This textbook uses a structured approach to learning Python while developing effective AI collaboration skills. Each chapter follows three distinct phases:

| Mode | Icon | AI Policy | Purpose |
|------|------|-----------|--------|
| **CRAWL** | 🐛 | No AI assistance | Build foundational skills you can demonstrate independently |
| **WALK** | 🚶 | AI for understanding only | Use AI to explain concepts and errors, but write your own code |
| **RUN** | 🚀 | Full AI collaboration | Partner with AI on complex tasks while documenting your process |

**Why This Matters:** Your exams will test CRAWL and WALK material with no AI assistance. If you skip the foundational work and rely entirely on AI, you won't pass. The progression ensures you build genuine competence before leveraging AI as a professional tool.

---

## What is a Jupyter Notebook?

Before we dive into Python, let's understand the tool you're using right now.

**A Jupyter notebook is an interactive document that combines:**
- Text and explanations (like this cell you're reading)
- Live code that you can run and modify
- Output from that code (numbers, graphs, tables, errors)
- Your own notes and observations

Think of it as a lab notebook for data analysis. Instead of separating your code files, documentation, and results into different places, everything lives together in one document. This makes it perfect for learning Python because you see explanations, try code immediately, and can document what you discover.

### Two Types of Cells

Every notebook contains two types of cells:

1. **Markdown cells** (like this one): Display formatted text, headers, lists, and images. Use these for explanations and documentation.

2. **Code cells**: Contain Python code you can execute. When you run a code cell, Python processes the commands and displays any output directly below the cell.

### How to Run Cells

To execute a cell:
- Click on it to select it (you'll see a blue or green border)
- Press **Shift + Enter** (runs the cell and moves to the next one)
- Or press **Ctrl + Enter** (runs the cell but stays on it)
- Or click the ▶️ "Run" button in the toolbar

Try it now with the code cell below. Click on it and press Shift + Enter:

## Environment Setup

Before running any code in this course, you need to make sure the required Python libraries are installed. You only need to do this **once** per environment.

**If you're using Anaconda:** These libraries are already included. You can skip the install step.

**If you're using pip (or aren't sure):** Run the cell below to install everything you need. If the packages are already installed, this will do nothing harmful — it will just confirm they're up to date.

In [80]:
# Run this cell once to install required libraries (safe to re-run)
# If you're using Anaconda, you can skip this step.
%pip install numpy pandas matplotlib seaborn --quiet

Note: you may need to restart the kernel to use updated packages.


In [1]:
# This is a code cell. Run it to see the output!
print("Welcome to BUAN 446: Python for Data Analysts")
print("You just executed your first Python code in a Jupyter notebook.")
print("Notice how the output appears right below this cell.")

Welcome to BUAN 446: Python for Data Analysts
You just executed your first Python code in a Jupyter notebook.
Notice how the output appears right below this cell.


Did you see the three messages appear below the code cell? That's how notebooks work: you write code, run it, see results, and continue.

### Working Through This Notebook

As you progress through this chapter, follow these practices:

- **Read every markdown cell.** They explain concepts and provide essential context.
- **Run every code cell.** Don't just read code, execute it and observe what happens.
- **Experiment freely.** Modify code cells and run them again. Making mistakes and fixing them is how you learn.
- **Add your own notes.** Create new markdown cells (Insert → Insert Cell Below) or add comments in code cells about what you're learning.

### Important Notes

**Cell execution order matters.** Notebooks typically run from top to bottom, but you can execute cells in any sequence. This flexibility is powerful but can cause confusion. If something stops working, try restarting the notebook: **Kernel → Restart & Run All** from the menu bar.

**Save your work.** Notebooks autosave periodically, but save manually anyway (Ctrl+S or Cmd+S). Your work is stored as a `.ipynb` file that you can share, email, or submit for assignments.

**You'll see numbered cells.** When you run a code cell, a number appears in brackets like `[1]` or `[2]`. This shows the execution order. If you see `[*]`, the cell is currently running.

Now you're ready to start learning Python.

## 📊 Course Case Study: Crestview Student Success Dataset

Throughout this course, we'll work with a synthetic dataset representing Crestview University student records. This dataset mirrors the kind of data you'd encounter in an institutional research or academic analytics role.

**The Dataset:** 600 student records with the following variables:

| Variable | Type | Description | Example |
|----------|------|-------------|----------|
| Student_ID | String | Unique identifier | CU100001 |
| College | String | Crestview college | "College of Business" |
| Major | String | Declared major | "Finance" |
| Class_Year | String | Academic standing | "Junior" |
| GPA | Float | Grade Point Average | 3.41 |
| Credits_Attempted | Integer | Total credits attempted | 105 |
| Credits_Earned | Integer | Credits successfully earned | 99 |

**Why This Dataset?**
- It's relevant to your experience as students
- It contains realistic patterns (GPA distributions, credit accumulation)
- It will evolve: starting clean, becoming messy, gaining new variables
- The same data will carry through all six weeks of this course

**The Journey:**
- **Weeks 1-2:** Work with the clean version to learn Python fundamentals
- **Week 3:** Confront a "messy" version with real-world data quality issues
- **Weeks 4-6:** Analyze the cleaned data using NumPy, Pandas, and visualization

This mirrors professional work: you rarely get perfect data. Learning to clean and validate data is as important as analyzing it.

## Learning Objectives

By the end of this chapter, you will:

- 🐛 Write basic Python expressions and statements from memory
- 🐛 Understand and use different data types (integers, floats, strings, booleans)
- 🐛 Create and manipulate variables with proper naming conventions
- 🐛 Perform arithmetic operations and understand operator precedence
- 🚶 Use AI tools to explain Python error messages
- 🚶 Format output using f-strings with AI-assisted learning
- 🚶 Convert between data types and handle conversion errors
- 🚀 Build a student record analyzer that calculates key academic metrics

---

# 🐛 CRAWL: Python Fundamentals

**Rules for this section:**
- Close all AI tools (ChatGPT, Claude, Copilot, etc.)
- Work through examples by typing them yourself
- Use only this notebook, Python documentation, or your instructor for help
- This material will appear on exams without AI assistance

---

## 📚 DataCamp Resources for Chapter 1

Crestview provides access to DataCamp to support your learning. The following courses and chapters align with this chapter's CRAWL material. Complete these **before or alongside** the exercises below to reinforce foundational concepts.

### Required DataCamp Content

**[Introduction to Python](https://www.datacamp.com/courses/intro-to-python-for-data-science)** - Complete these chapters:

| Chapter | Topics Covered | Alignment |
|---------|---------------|------------|
| Chapter 1: Python Basics | Variables, types, arithmetic operators | Sections 1.1-1.4 of this chapter |
| Chapter 2: Python Lists | List creation, indexing, manipulation | Preview for Chapter 2 |

**Estimated time:** 2-3 hours total

### How to Use DataCamp

DataCamp courses are interactive tutorials where you type code and get immediate feedback. This is **allowed** in CRAWL mode because:

1. You're writing the code yourself (not asking AI to generate it)
2. The exercises test your understanding with immediate correction
3. The guided format builds muscle memory for syntax

**Workflow recommendation:**
1. Complete the DataCamp chapter first to learn the concepts interactively
2. Return to this notebook and attempt the practice problems without looking back
3. If stuck, re-read the relevant section in this notebook (not DataCamp)

### Optional Supplementary Content

If you want additional practice beyond the required content:

- **[Data Types for Data Science in Python](https://www.datacamp.com/courses/data-types-for-data-science-in-python)** - Chapter 1 covers fundamental types in more depth
- **[Python Fundamentals](https://www.datacamp.com/courses/python-fundamentals)** - Reinforces basics with different examples

---

## 1.1 Your First Python Code

Python executes code line by line. The simplest thing you can do is display output using the `print()` function.

In [82]:
print("Hello, Data Analyst!")

Hello, Data Analyst!


Run the cell above (Shift+Enter). You should see the text displayed below it.

Python can also evaluate expressions directly. In Jupyter notebooks, the last expression in a cell automatically displays its value:

In [83]:
2 + 3

5

In [84]:
2 + 3
10 * 5

50

## 1.2 Variables and Assignment

A **variable** is a name that refers to a value stored in memory. You create variables using the assignment operator `=`.

We also see an example of a **comment** below. `#` denotes a comment and python ignores anything on a line after a `#`

Let's start working with our student dataset context:

In [85]:
# Creating variables for a student record and assigning values to variables
student_id = "CU100001"
gpa = 3.41
credits_attempted = 105
credits_earned = 99

# Using variables to calculate
credits_failed = credits_attempted - credits_earned
print(credits_failed)

6


### Variable Naming Rules

Python has strict rules about variable names:

1. Must start with a letter or underscore (`_`)
2. Can contain letters, numbers, and underscores
3. Cannot be a Python keyword (like `if`, `for`, `while`, `print`)
4. Case-sensitive (`GPA` and `gpa` are different variables)

**Best Practice:** Use `snake_case` for variable names (lowercase with underscores).

In [86]:
# Good variable names for student data
total_students = 600
avg_gpa = 3.27
college_name = "College of Business"

print(total_students)
print(avg_gpa)

600
3.27


In [87]:
# This will cause an error - run it to see what happens
first_year_count = 135

## 1.3 Data Types

Every value in Python has a **type**. The four fundamental types you need to know:

| Type | Python Name | Examples | Student Data Use Case |
|------|-------------|----------|----------|
| Integer | `int` | `42`, `-7`, `0` | Credits, student counts |
| Float | `float` | `3.14`, `-0.5`, `2.0` | GPA, completion rates |
| String | `str` | `"hello"`, `'data'` | Names, IDs, colleges |
| Boolean | `bool` | `True`, `False` | On probation? Graduated? |

### A Quick Look at Functions

You've already seen `print(...)`, which is a **function**. In Python, a function is a reusable piece of code that performs a specific task. 

Key things to know about functions:
1. **Name**: Every function has a name (like `print`, `input`, or `type`).
2. **Parentheses**: We use `()` to "call" (run) the function.
3. **Arguments**: We put data inside the parentheses for the function to use.

For example, when you write `print("Hello")`, you are calling the `print` function and passing `"Hello"` as an argument.

Now, let's use a new function called `type()`.

Use the `type()` function to check a value's type:

In [88]:
print(type(42))
print(type(3.14))
print(type("hello"))
print(type(True))

<class 'int'>
<class 'float'>
<class 'str'>
<class 'bool'>


In [89]:
# Variables in our student dataset and their types
student_id = "CU100001"     # str - text identifier
gpa = 3.41                  # float - decimal number
credits_earned = 99         # int - whole number
is_graduate = False         # bool - true/false flag

print(type(student_id))
print(type(gpa))
print(type(credits_earned))
print(type(is_graduate))

<class 'str'>
<class 'float'>
<class 'int'>
<class 'bool'>


### Important Type Behaviors & Dynamic Typing

Python is a **dynamically typed** language. This means:

1. **Variables don't have a fixed type**: You can assign an integer to a variable `x` now, and a string to `x` later (though it's often confusing to do so).
2. **Types are attached to values, not variables**: `10` is an integer, `"10"` is a string. The variable just points to the value.
3. **Operations depend on the value's type**: The meaning of `+` changes depending on whether you are adding numbers or strings.

Let's see how different types behave differently with the same operators:

In [90]:
# Addition with numbers
print(105 + 15)        # Mathematical addition: total credits

# Addition with strings (concatenation)
print("CU" + "100001")    # String joining: building an ID

120
CU100001


In [91]:
# Multiplication with numbers
print(3 * 15)   # 3 semesters x 15 credits each

# Multiplication with strings (repetition)
print("=" * 40)  # Create a divider line by repeating the provided string 40 times

45


In [92]:
# This will cause an error - you cannot add a string and an integer
print("Credits: " + "105")

Credits: 105


## 1.4 Arithmetic Operators

Python supports standard mathematical operations:

| Operator | Operation | Example | Result |
|----------|-----------|---------|--------|
| `+` | Addition | `5 + 3` | `8` |
| `-` | Subtraction | `5 - 3` | `2` |
| `*` | Multiplication | `5 * 3` | `15` |
| `/` | Division | `5 / 3` | `1.666...` |
| `//` | Floor Division | `5 // 3` | `1` |
| `%` | Modulo (remainder) | `5 % 3` | `2` |
| `**` | Exponentiation | `5 ** 3` | `125` |

### Understanding Floor Division (`//`) and Modulo (`%`)

These two operators are often new to beginners but are extremely useful in data analysis.

**Floor Division (`//`)**:
- Also called "integer division".
- It divides two numbers and rounds the result *down* to the nearest whole number.
- **Use case**: Checking how many full groups fit into a total.
  - *Example*: If you have 105 students and want to form groups of 4, `105 // 4` tells you how many full groups you can make.

**Modulo (`%`)**:
- Returns the **remainder** of a division.
- **Use case**: Finding what's left over, or checking for patterns.
  - *Example*: `105 % 4` tells you how many students are left without a group.
  - *Example*: `number % 2` is a standard way to check if a number is even (result is 0) or odd (result is 1).

In [93]:
# Division always returns a float
print(100 / 4)       # 25.0, not 25

# Floor division truncates to integer
print(105 // 15)     # How many full semesters of 15 credits?

# Modulo gives the remainder as an integer
print(105 % 15)      # Credits left over after full semesters

25.0
7
0


In [94]:
# Try it yourself: what code could you write to confirm that division always returns a float?


In [95]:
# Practical example: calculating credit completion rate
credits_attempted = 105
credits_earned = 99

completion_rate = credits_earned / credits_attempted
print(completion_rate)

# As a percentage
completion_percentage = (credits_earned / credits_attempted) * 100
print(completion_percentage)

0.9428571428571428
94.28571428571428


In [96]:
# Calculating average GPA across students
total_gpa_points = 3.41 + 3.55 + 3.60 + 3.44 + 3.90
student_count = 5

average_gpa = total_gpa_points / student_count
print(average_gpa)

3.5799999999999996


### Operator Precedence

Python follows standard mathematical order of operations (PEMDAS):

1. Parentheses `()`
2. Exponentiation `**`
3. Multiplication, Division, Floor Division, Modulo `* / // %`
4. Addition, Subtraction `+ -`

When in doubt, use parentheses to make your intent clear.

In [97]:
# Without parentheses
result1 = 2 + 3 * 4 # could be written 2 + (3 * 4) for clarity
print(result1)  # 14, not 20

# With parentheses if you want to add the numbers before multiplying by 4
result2 = (2 + 3) * 4
print(result2)  # 20

14
20


In [98]:
# GPA calculation example
# Quality points = credits * grade points
# GPA = total quality points / total credits

credits_course1 = 3
grade_course1 = 4.0  # A

credits_course2 = 4
grade_course2 = 3.0  # B

# This needs parentheses to work correctly!
gpa = (credits_course1 * grade_course1 + credits_course2 * grade_course2) / (credits_course1 + credits_course2)
print(gpa)

3.4285714285714284


## 1.5 Getting User Input

The `input()` function pauses execution and waits for the user to type something. It always returns a string.

### Why Do We Need User Input?

Up to this point, all our variables have been hard-coded (e.g., `gpa = 3.41`). While this is fine for testing, real-world tools need to be flexible. 

Using `input()` allows your code to:
1. **Be Interactive**: The program can ask questions and respond to answers.
2. **Be Dynamic**: The same script can process different data each time runs.
3. **Automate Workflows**: You can write a script once (e.g., "Calculate Graduation Eligibility") and use it for any student by simply entering their ID or credits when prompted.

In [99]:
student_name = input("Enter student name: ")
print("Looking up records for:", student_name)

Enter student name:  Evan


Looking up records for: Evan


In [100]:
# Important: input() returns a string, even if you type a number
credits = input("Enter credits attempted: ")
print(type(credits))  # <class 'str'> - it's text, not a number!

Enter credits attempted:  40


<class 'str'>


---

## 🐛 CRAWL Practice Problems

Complete these problems without any AI assistance. Write your code in the cells provided.

**Context:** These problems use the Crestview student dataset scenario. Imagine you're building tools for an academic advisor or institutional researcher.

---

### Problem 1.1: Student Record Variables
Create variables for a single student record:
- Student ID: "CU100042"
- College: "College of Business"
- GPA: 3.75
- Credits attempted: 89
- Credits earned: 85

Then calculate and print how many credits this student did not earn.

In [101]:
# Your code here
student_id = "CU100042"
college = "College of Business"
gpa = 3.75
credits_attempted = 89
credits_earned = 85

print(f"Unearned credits: {credits_attempted-credits_earned}")

Unearned credits: 4


---> **Instructor Feedback — Problem 1.1: Full Credit ✅**>> Correct. All five variables are assigned with the right types, and you compute and print the unearned credits (4) with a clear, labeled f-string. Good habit naming the output.

### Problem 1.2: Credit Distribution
A student has attempted 105 credits. Assuming a typical semester is 15 credits:
- How many complete semesters of 15 credits have they taken?
- How many extra credits beyond those complete semesters?

Use floor division and modulo to calculate these values.

In [102]:
# Your code here
attempts = 105
credits_per_semester = 15

complete_semesters = 105 // 15
extra_beyond = 105 % 15
print(complete_semesters, extra_beyond)



7 0


---> **Instructor Feedback — Problem 1.2: Full Credit ✅**>> Correct output: 7 complete semesters and 0 leftover credits, using floor division and modulo. One habit to tighten: you created `attempts = 105` and `credits_per_semester = 15` but then hardcoded `105 // 15` in the calculation. Reference the variables (`attempts // credits_per_semester`) so the logic updates automatically if the numbers change.

### Problem 1.3: Predict the Output
What is the output of each expression? Predict first, then run to check.

```python
a) 120 // 15      # Semesters from 120 credits
b) 99 / 105       # Completion rate (as decimal)
c) 3.5 + 3.8 + 4.0 / 3   # Average GPA? Or is it?
d) (3.5 + 3.8 + 4.0) / 3  # Now what?
```

In [103]:
# Check your predictions here
# a - 8, b = .942, c = 8.6, d 3.8
print(120//15)
print(99/105)
print(3.5 + 3.8 + 4.0 / 3)
print((3.5+3.8+4.0)/3)

8
0.9428571428571428
8.633333333333333
3.766666666666667


---> **Instructor Feedback — Problem 1.3: Full Credit ✅**>> Well done. You predicted each result in the comment first and then ran all four to check, which is exactly the intended workflow. Your estimates for c (8.6) and d (3.8) were close. Note the precedence difference you surfaced: c evaluates `4.0/3` before adding (8.63), while d averages all three values inside the parentheses (3.77).

### Problem 1.4: Interactive Student Lookup
Write code that:
1. Asks the user for a student ID
2. Asks the user for the student's major
3. Prints a message like: "Student CU100042 is majoring in Finance"

##### Your code here
student_id = input("Please Enter Student ID: ")
major = input("Please Input Student Major: ")
print(f"Student {student_id} is majoring in {major}")

---> **Instructor Feedback — Problem 1.4: Full Credit ✅**>> Logic is correct: you prompt for the ID and major and format the output exactly as asked. One fix for next time: this answer is in a **markdown cell**, so it renders as text and never actually executes. Select the cell, press Esc then Y to convert it to a code cell so it runs.

### Problem 1.5: College Statistics
The Crestview student dataset has this distribution across colleges:
- College of Business: 109 students
- Engineering: 114 students  
- Arts and Sciences: 121 students
- College of Health: 135 students
- College of Education: 121 students

Calculate and print:
1. The total number of students
2. The percentage of students in the College of Health (to 1 decimal place, you'll learn formatting soon but for now just print the raw number)

In [105]:
# Your code here
business = 109
eng = 114
arts = 121
health = 135
ed = 121

print(f"Total Students: {business + eng + arts + health + ed}")
print(f"Percent Health: {health / (business + eng + arts + health + ed):.1f}")

Total Students: 600
Percent Health: 0.2


---> **Instructor Feedback — Problem 1.5: No Credit ⚠️**>> The total is right (600), but the Health percentage is not. You divided `health / total` **without multiplying by 100**, then formatted to one decimal, so the output prints `0.2` instead of the `22.5` percent the problem asks for. Multiply the ratio by 100: `health / total * 100`.

---

# 🚶 WALK: Building Proficiency

**Rules for this section:**
- You may use AI tools to **explain** concepts, error messages, and documentation
- You must **write all code yourself** after understanding it
- Do NOT ask AI to write the solution for you
- Good prompts: "Explain what this error means" or "What does the `:.2f` in an f-string do?"
- Bad prompts: "Write code that does X" or "Fix my code"

---

## 1.6 Type Conversion

You can convert between types using conversion functions:

| Function | Converts to | Example |
|----------|-------------|--------|
| `int()` | Integer | `int("42")` → `42` |
| `float()` | Float | `float("3.14")` → `3.14` |
| `str()` | String | `str(42)` → `"42"` |
| `bool()` | Boolean | `bool(0)` → `False` |

In [106]:
# Converting input to a number for calculations
credits_str = input("Enter credits earned: ")
credits = int(credits_str)
semesters = credits // 15
print("Approximate semesters completed:", semesters)

Enter credits earned:  150


Approximate semesters completed: 10


> **A Quick Preview:** The code below uses an `if` statement to check whether the GPA qualifies for the Dean's List. We haven't covered `if` statements yet — that's coming in **Chapter 2: Control Flow**, where you'll learn how to make your programs respond differently based on conditions. For now, just notice how readable the logic is: *if the GPA is at least 3.5, print a message.* You'll be writing your own conditional logic soon!

In [107]:
# Combining conversion with input in one line
gpa = float(input("Enter GPA: "))
if gpa >= 3.5:
    print("Dean's List eligible!")

Enter GPA:  3.9


Dean's List eligible!


### When Conversions Fail

Not all conversions are valid. Run the cell below to see what happens:

In [109]:
# This will cause a ValueError
gpa = 3.5 #float("three point five")

**AI Learning Opportunity:** Copy that error message and ask an AI to explain what it means and why it happened. This is exactly the kind of understanding-focused AI use that will help you learn.

## 1.7 F-Strings (Formatted String Literals)

F-strings provide a clean way to embed expressions inside strings. Start the string with `f` and put expressions inside `{}`.

**Reference:** For a comprehensive guide, check out [f-string help](https://fstring.help/).

In [110]:
student_id = "CU100001"
gpa = 3.41
college = "College of Business"

# Without f-string (awkward)
# Remember when we use `+` with strings it concatenates (puts them together)
print("Student " + student_id + " has a GPA of " + str(gpa))

# With f-string (clean)
# Just one set of quotes at the beginning and end of the string and variable are in curly braces
print(f"Student {student_id} has a GPA of {gpa}")

Student CU100001 has a GPA of 3.41
Student CU100001 has a GPA of 3.41


### Format Specifiers

F-strings support format specifiers after a colon inside the braces. These are essential for professional data reporting and another advantage of f-strings:

In [111]:
credits_attempted = 105
credits_earned = 99
completion_rate = credits_earned / credits_attempted

# .2f means "2 decimal places, float format"
print(f"Credits earned: {credits_earned}")
print(f"Completion rate: {completion_rate:.2f}")  # 0.94
print(f"Completion rate: {completion_rate:.1%}")  # 94.3% (percentage format!)

Credits earned: 99
Completion rate: 0.94
Completion rate: 94.3%


In [112]:
# Formatting student data for reports
gpa = 3.4142857
total_students = 600

print(f"Average GPA: {gpa:.2f}")           # 3.41
print(f"Total enrollment: {total_students:,}")  # 600 (would show 1,234 for larger numbers - try it out)
print(f"GPA (3 decimals): {gpa:.3f}")      # 3.414

Average GPA: 3.41
Total enrollment: 600
GPA (3 decimals): 3.414


In [113]:
# Creating a formatted student report
student_id = "CU100001"
major = "Accounting"
gpa = 3.41
credits = 105

print("=" * 40)
print("STUDENT RECORD")
print("=" * 40)
print(f"ID:      {student_id}")
print(f"Major:   {major}")
print(f"GPA:     {gpa:.2f}")
print(f"Credits: {credits}")
print("=" * 40)

STUDENT RECORD
ID:      CU100001
Major:   Accounting
GPA:     3.41
Credits: 105


**AI Learning Opportunity:** The format specifier syntax has many options. If you want to do something specific (like align text in columns or pad with zeros), ask an AI: "How do I right-align text in a Python f-string to create columns?"

## 1.8 Comments and Code Documentation

Comments explain your code to others (and to your future self). Python ignores everything after `#`.

In [114]:
# Calculate credit completion rate for a student
# This metric helps identify students who may be struggling

credits_attempted = 105    # Total credits registered for
credits_earned = 99        # Credits successfully passed

# Calculate completion rate (earned / attempted)
completion_rate = credits_earned / credits_attempted

# Convert to percentage for readability
completion_percentage = completion_rate * 100

print(f"Completion rate: {completion_percentage:.1f}%")

Completion rate: 94.3%


### Comment Best Practices

Good comments explain **why**, not **what**. The code already shows what it does.

In [115]:
# Bad comment (just restates the code)
gpa = gpa + 0.1  # Add 0.1 to gpa

# Good comment (explains purpose)
gpa = gpa + 0.1  # Grade replacement policy allows one course improvement

---

## 🚶 WALK Practice Problems

Use AI to help you understand concepts and errors, but write all code yourself.

---

### Problem 1.6: GPA Calculator
Write a program that:
1. Asks the user for their current GPA
2. Asks how many credit hours they've completed
3. Calculates and displays how many quality points they have (GPA × credits)
4. Format the output to 2 decimal places

If you get stuck on formatting, ask AI to explain (not to write the code).

In [116]:
# Your code here
gpa = float(input("What's your current GPA: "))
hours = int(input("How Many Credit Hours Have You Completed: "))
quality_points = gpa * hours
print(f"Quality Points: {quality_points:0.2f}")

What's your current GPA:  2.9
How Many Credit Hours Have You Completed:  80


Quality Points: 232.00


---> **Instructor Feedback — Problem 1.6: Full Credit ✅**>> Correct. GPA and credit hours are captured as numbers, quality points are computed (232.00), and the output is formatted to two decimals. `:0.2f` and `:.2f` behave the same here.

### Problem 1.7: Student Report Card
Create a formatted report for a student with this information:
- Student ID: CU100099
- Name: Maria Santos
- Major: Computer Science
- Class Year: Junior
- GPA: 3.8571428 (display as 3.86)
- Credits Attempted: 89
- Credits Earned: 86
- Completion Rate: (calculate and show as percentage)

Make it look professional with headers and alignment.

In [117]:
# Your code here
sid = "CU100099"
name = "Mario Santos"
major = "Computer Science"
year = "Junior"
gpa = 3.8571428
credits_attempted = 89
credits_earned = 86
completion_rate = credits_earned / credits_attempted
print(f"|{"-"*41}|")
print(f"|{"Student":^20}|{"Value":^20}|")
print(f"|{"-"*41}|")
print(f"|{"Student ID":^20}|{sid:^20}|")
print(f"|{"Name":^20}|{name:^20}|")
print(f"|{"Major":^20}|{major:^20}|")
print(f"|{"Class Year":^20}|{year:^20}|")
print(f"|{"GPA":^20}|{gpa:^20,.2f}|")
print(f"|{"Credits Attempted":^20}|{credits_attempted:^20}|")
print(f"|{"Credits Earned":^20}|{credits_earned:^20}|")
print(f"|{"Completion Rate":^20}|{completion_rate:^20,.2%}|")
print(f"|{"-"*41}|")

|-----------------------------------------|
|      Student       |       Value        |
|-----------------------------------------|
|     Student ID     |      CU100099      |
|        Name        |    Mario Santos    |
|       Major        |  Computer Science  |
|     Class Year     |       Junior       |
|        GPA         |        3.86        |
| Credits Attempted  |         89         |
|   Credits Earned   |         86         |
|  Completion Rate   |       96.63%       |
|-----------------------------------------|


---> **Instructor Feedback — Problem 1.7: Full Credit ✅**>> Strong, professional table with the correct GPA (3.86) and completion rate shown as a percentage. Three cleanups: the name should be **"Maria Santos"** (you entered "Mario"); the `"-"*41` divider is narrower than the table body, so the top/bottom borders don't line up with the columns; and the same-quote nested f-strings such as `f"|{"-"*41}|"` only run on **Python 3.12+** and raise a SyntaxError on older kernels. Use single quotes inside the braces for portability.

### Problem 1.8: Debug These Errors
Run the following code cells. Each has an error. Use AI to help you understand the error messages, then fix them yourself.

In [118]:
# Error 1: Fix this code
total_credits = int("105") + 15

In [119]:
# Error 2: Fix this code
gpa = float(input("Enter GPA: "))  # User enters "3.5"
print(f"Your GPA is {gpa}")

Enter GPA:  3.5


Your GPA is 3.5


In [120]:
# Error 3: Fix this code
class_year = "Junior"
print(class_year)

Junior


---> **Instructor Feedback — Problem 1.8: Full Credit ✅**>> All three errors fixed correctly: the string-to-int conversion, the missing closing parenthesis on `input()`, and the invalid `class-year` identifier renamed to `class_year`.

---

# 🚀 RUN: Real-World Application

**Rules for this section:**
- Full AI collaboration is encouraged
- Document your process: what prompts you used, what AI suggested, what you modified
- You must understand the final solution and be able to explain every line
- Add comments showing your understanding

---

## Chapter Project: Student Academic Metrics Calculator

Build an interactive tool that calculates key academic metrics for a student. This mirrors what an academic advisor might use.

### Requirements

**The program should:**
1. Collect student information via input:
   - Student ID
   - Current GPA
   - Credits attempted
   - Credits earned

2. Calculate and display:
   - Credit completion rate (as percentage)
   - Quality points (GPA × credits earned)
   - Approximate semesters completed (credits ÷ 15)
   - Credits needed to graduate (assuming 120 required)

3. Generate a formatted report that looks professional

**Bonus challenges:**
- Add Dean's List status (GPA ≥ 3.5 with at least 12 credits)
- Calculate what GPA is needed in remaining credits to reach 3.0 or 3.5 overall
- Handle the case where the student has already met graduation requirements

### Working with AI

Effective prompts to consider:
- "How do I format numbers as percentages in Python f-strings?"
- "What's the formula to calculate required GPA in remaining credits to achieve a target GPA?"
- "How can I make my printed output align in columns?"

Avoid prompts like:
- "Write a student metrics calculator for me"
- "Complete this assignment"

The goal is to use AI as a knowledgeable colleague, not as someone who does your work.

In [ ]:
# STUDENT ACADEMIC METRICS CALCULATOR
# 
# Document your AI collaboration below:
# - Prompts used: in a jupyter notebook, using python, i want to input multiple students worth of data.  I have an input("Do you want to do more y/n") at the end, is this the best way to handle it in the notebook framework, is there something cute that jupyter would interpret as a button for yes/no?
# - Key insights from AI: the "ipywidgets" library exists and that's super cool, but i'm not going to do that right now...
# - Modifications you made: none, though i'm going to explore those ipywidgets features... that's freaking cool
#
# Your code below:

#some magic numbers so we don't have to dig if requirements change
TOTAL_CREDITS = 120
DEANS_LIST_GPA = 3.5

#the future gpa is really a quality points question
REQUIRED_QUALITY_POINTS = TOTAL_CREDITS * DEANS_LIST_GPA


print("=" * 80)
print("CRESTVIEW UNIVERSITY - STUDENT METRICS CALCULATOR")
print("=" * 80)

#array to keep track of all the students we're going to enter later
students = []


#easy dict to allow for column renaming if something changes or we want to make it shorter
col_key_to_name = {
    'id': 'Student ID',
    'gpa': 'GPA',
    'cr_attempted': 'Credits Attempted',
    'cr_earned': 'Credits Earned',
    'completion_rate': 'Comp. Rate',
    'quality_points': 'Quality Points',
    'semesters_complete': 'Semesters Complete',
    'credits_remain':'Credits Remaining',
    'deans_list': 'Deans List',
    'req_gpa_remain':'Required GPA'
}

#padding separate from column length is a design decision, but i like it.
PADDING = 2

#this keeps track of the column widths for all the fields.  A smarter version of this would loop through the data to see if any 
#data in the fields would require a wider entry, but we're not there right now.  The headers are longer than the data, won't always be true
# (see gpa below)
col_key_to_width = {
    'id': len(col_key_to_name['id']),
    'gpa': len(col_key_to_name['gpa']) + 1, #gpas are 4 characters and i don't want to dynamically size columns on the first assignment
    'cr_attempted': len(col_key_to_name['cr_attempted']),
    'cr_earned': len(col_key_to_name['cr_earned']),
    'completion_rate': len(col_key_to_name['completion_rate']),
    'quality_points': len(col_key_to_name['quality_points']),
    'semesters_complete': len(col_key_to_name['semesters_complete']),
    'credits_remain':len(col_key_to_name['credits_remain']),
    'deans_list': len(col_key_to_name['deans_list']),
    'req_gpa_remain' :len(col_key_to_name['req_gpa_remain']),
}

#loop until break
while True:
    #new dict for every lovely student
    student = {}
    student['id'] = input("Enter Student ID: ").upper() #so that su -> SU for consistency
    student['gpa'] = float(input("Enter GPA: "))
    student['cr_attempted'] = int(input("Enter Credits Attempted: "))
    student['cr_earned'] = int(input("Enter Credits Earned: "))
    student['completion_rate'] = student['cr_earned'] / student['cr_attempted']
    student['quality_points'] = student['gpa'] * student['cr_earned'] 
    student['semesters_complete'] = student['cr_earned'] // 15
    student['credits_remain'] = max(TOTAL_CREDITS - student['cr_earned'],0)
    student['deans_list'] = student['gpa'] >= 3.5 and student['cr_earned'] >= 12


    #this is a bad thing in general but given this is a report, i'm actually going to stringify these
    #now, that way I can just loop print everythign later and it'll work nice, also allows me to dynamically 
    #calculate the required field width if i wanted to do that...
    
    
    
    #Do this first as a string to cut things out
    if student['credits_remain'] != 0:
        student['req_gpa_remain'] = (REQUIRED_QUALITY_POINTS - student['quality_points']) / student['credits_remain']
        if student['req_gpa_remain'] > 4.0 or student['req_gpa_remain'] < 0:
            student['req_gpa_remain'] = 'N/A'
        else:
            student['req_gpa_remain'] = f"{student['req_gpa_remain']:.2f}"
    else: 
        student['req_gpa_remain'] = 'N/A'

    #also nice because we only have to format here.
    student['gpa'] = f"{student['gpa']:.2f}" 
    student['cr_attempted'] = f"{student['cr_attempted']}" 
    student['cr_earned'] = f"{student['cr_earned']}" 
    student['completion_rate'] = f"{student['completion_rate']:.2%}" 
    student['quality_points'] = f"{student['quality_points']:.2f}" 
    student['semesters_complete'] = f"{student['semesters_complete']}" 
    student['credits_remain'] = f"{student['credits_remain']}" 
    if student['deans_list']:
        student['deans_list'] = "✓"
    else:
        student['deans_list'] = "✗"
    
    students.append(student)
    if input("More entries? [y/N]: ").lower() != 'y':
        break

#let it breath!
print()
#print headers
printed_width = 0
for key,header in col_key_to_name.items():
    print(f"|{header:^{col_key_to_width[key]+PADDING}}",end='')
    printed_width += col_key_to_width[key]+PADDING + 1 #extra 1 for the '|'
print("|")
printed_width+=1 #another |


print('-' * printed_width)

#the magic, dicts have fixed order, so we know it will always match,
#nested format strings allow us to dynamically size the fields based
#on the dict above
for student in students:
    for field,value in student.items():
        print(f"|{value:^{col_key_to_width[field]+PADDING}}",end='')
    print("|")


print('-' * printed_width)    




---> **Instructor Feedback — Chapter Project: Works (100%) = 30 pts ✅**>> Excellent work that goes well beyond the brief. The multi-student entry loop, dictionary-driven dynamic column widths, and **all three bonus challenges** are present and working: Dean's List status, required GPA in remaining credits (with sensible N/A handling for out-of-range values), and the graduation-met case via `max(TOTAL_CREDITS - earned, 0)`. Your AI collaboration notes are specific and honest. This is the standard to aim for.

### Project Reflection

After completing the project, answer these questions in the cell below:

1. What was the most challenging part of this project?
2. What did you learn from your AI collaboration that you didn't know before?
3. Is there any part of the code you couldn't explain if asked? If so, review it now.
4. How does this connect to the Crestview student dataset we'll be working with throughout the course?

*Your reflection here:*
1. Not super challenging, I would have liked to do a better job on teh output format.  I think that moving to 2 lines could condense the width of the report output, but I didn't have time
2. There's a whole amazing library of things that jupyter notebook can utilize for richer input/output, that's sweet!
3. I got this...
4. The dataset is something that we're going to injest and analyze, obviously we won't be using input, we'll be dumping things in line by line and processing them, then running analytics on all of it in aggregate.


---

# Preview: What's Coming

In **Chapter 2**, you'll learn control flow (if/else, loops) which will let you:
- Classify students by academic standing
- Identify students at risk (GPA < 2.0, low completion rate)
- Process multiple student records
- Calculate statistics across groups

In **Week 3**, you'll encounter the **messy version** of this dataset with real-world data quality issues:
- Inconsistent college names ("COB" vs "College of Business")
- Missing GPA values
- Typos in majors
- Duplicate records

The skills you're building now are the foundation for handling those challenges.

---

# Accountability Check

Before moving to Chapter 2, honestly assess yourself:

## 🐛 CRAWL Competencies (Must be able to do without notes or AI)
- [x] Create variables with appropriate names and types
- [x] Perform arithmetic operations and predict results
- [x] Use `print()` to display output
- [x] Use `input()` to get user data
- [x] Identify the type of any value using `type()`
- [x] Calculate completion rate: `earned / attempted`

## 🚶 WALK Competencies (Can use AI to learn, must write code yourself)
- [x] Convert between types using `int()`, `float()`, `str()`
- [x] Use f-strings with format specifiers (`.2f`, `.1%`)
- [x] Read and understand Python error messages
- [x] Write meaningful comments that explain *why*

## 🚀 RUN Competencies (AI-assisted, must understand completely)
- [x] Build multi-step programs that process user input
- [x] Format output professionally for reports
- [x] Document your code and AI collaboration process

**If you cannot check all the CRAWL boxes from memory, review before proceeding.**

---> **Instructor Feedback — Project Reflection: Not Completed ⚠️**>> The four reflection questions are unanswered. Please complete them. The reflection is a required component and reinforces the reasoning behind the project.